exercises/src/test/scala/datastructures/ListSpec.scala

```diff
@@ -0,0 +1,417 @@
+      List.append2(List(1,2,3),List(4,5)) mustEqual List(1,2,3,4,5)
+      List.append2(List('a','b','c'),List('d')) mustEqual List('a','b','c','d')
+    }
+  }
+
+  //========================================================================================
+  //#3.15
+  "concatenate" should {
+    "result in Nil for two empty Lists" in {
+      List.concatenate(List(), List()) mustEqual Nil
+    }
+
+    "return the other list when another List is Nil" in {
+      List.concatenate(List(1),List()) mustEqual List(1)
+      List.concatenate(List(),List('a')) mustEqual List('a')
+    }
+
+    "reslut in a List with multiple-elements with first list on the left and second on the right - two lists" in {
+      List.concatenate(List(1,2,3),List(4,5)) mustEqual List(1,2,3,4,5)
+      List.concatenate(List('a','b','c'),List('d')) mustEqual List('a','b','c','d')
+    }
+
+    "reslut in a List with multiple-elements with first list on the left and second in the middle and third on the right - three lists" in {
+      List.concatenate(List(1,2,3),List(4,5),List(6)) mustEqual List(1,2,3,4,5,6)
+      List.concatenate(List('a','b','c'),List('d','e'),List('f')) mustEqual List('a','b','c','d','e','f')
+    }
+  }
+
+  //========================================================================================
+  //#3.16
+  "add1" should {
+    "result in Nil for an empty List" in {
+      List.add1(List()) mustEqual Nil
+    }
+
+    "return a List with 1 added to a single-element List" in {
+      List.add1(List(1)) mustEqual List(2)
+    }
+
+    "return a List with 1 added to all elements in a multiple-element List" in {
+      List.add1(List(1,2,3)) mustEqual List(2,3,4)
+    }
+  }
+
+  //========================================================================================
+  //#3.17
+  "listDblToListStr" should {
+    "result in Nil for an empty List" in {
+      List.listDblToListStr(List()) mustEqual Nil
+    }
+
+    "return a List of String for a single-element List of Double" in {
+      List.listDblToListStr(List(1.0)) mustEqual Cons("1.0",Nil)
+    }
+
+    "return a List of String for a multiple-element List of Double" in {
+      List.listDblToListStr(List(1.1,2.2)) mustEqual Cons("1.1", Cons("2.2",Nil))
+    }
+  }
+
+  //========================================================================================
+  //#3.18
+  "map" should {
+    "result in Nil for an empty List" in {
+      List.map(List():List[Int])(_+1) mustEqual Nil
+    }
+
+    "apply the fucntion to the single-element List" in {
+      List.map(List(1))(_+1) mustEqual List(2)
+    }
+
+    "apply the function to all elements in multiple-element List" in {
+      List.map(List(1,2))(_*2) mustEqual List(2,4)
+    }
+  }
+
+  //========================================================================================
+  //#3.19
+  "filter" should {
+    "result in Nil for an empty List" in {
+      List.filter(List():List[Int])(_>0) mustEqual Nil
```

Comment:
You can also explicitly pass in a type parameter to help Scala's type inference and thus avoid having to annotate the type later: `List[Int]()` instead of `List(): List[Int]`.